govcd/access_control.go

```diff
@@ -0,0 +1,230 @@
+// with accessControl.IsSharedToEveryone = false, accessControl.AccessSettings defines which subjects can access the vApp
+// For each setting we must provide:
+// * The subject (HREF and Type are mandatory)
+// * The access level (one of ReadOnly, Change, FullControl)
+func (client *Client) SetAccessControl(accessControl *types.ControlAccessParams, href, entityType, entityName string) error {
+
+	href += "/action/controlAccess"
+	// Make sure that subjects in the setting list are used only once
+	if accessControl.AccessSettings != nil && len(accessControl.AccessSettings.AccessSetting) > 0 {
+		if accessControl.IsSharedToEveryone {
+			return fmt.Errorf("[client.SetAccessControl] can't set IsSharedToEveryone and AccessSettings at the same time for %s %s (%s)", entityType, entityName, href)
+		}
+		var used = make(map[string]bool)
+		for _, setting := range accessControl.AccessSettings.AccessSetting {
+			_, seen := used[setting.Subject.HREF]
+			if seen {
+				return fmt.Errorf("[client.SetAccessControl] subject %s (%s) used more than once", setting.Subject.Name, setting.Subject.HREF)
+			}
+			used[setting.Subject.HREF] = true
+			if setting.Subject.Type == "" {
+				return fmt.Errorf("[client.SetAccessControl] subject %s (%s) has no type defined", setting.Subject.Name, setting.Subject.HREF)
+			}
+		}
+	}
+	errorMessage := fmt.Sprintf("[client.SetAccessControl] error setting access control for %s %s (%s)", entityType, entityName, href)
+
+	accessControl.Xmlns = types.XMLNamespaceVCloud
+	resp, err := client.ExecuteRequest(href, http.MethodPost, types.MimeControlAccess,
+		errorMessage+": %s", accessControl, nil)
+
+	if err != nil {
+		return err
+	}
+	if resp == nil {
+		return fmt.Errorf("[client.SetAccessControl] nil response received")
+	}
+	return nil
+}
+
+// GetAccessControl retrieves the access control information for this vApp
+func (vapp VApp) GetAccessControl() (*types.ControlAccessParams, error) {
+
+	if vapp.VApp.HREF == "" {
+		return nil, fmt.Errorf("vApp HREF is empty")
+	}
+
+	return vapp.client.GetAccessControl(vapp.VApp.HREF, "vApp", vapp.VApp.Name)
+}
+
+// SetAccessControl changes the access control information for this vApp
+func (vapp VApp) SetAccessControl(accessControl *types.ControlAccessParams) error {
+
+	if vapp.VApp.HREF == "" {
+		return fmt.Errorf("vApp HREF is empty")
+	}
+
+	return vapp.client.SetAccessControl(accessControl, vapp.VApp.HREF, "vApp", vapp.VApp.Name)
+
+}
+
+// RemoveAccessControl is a shortcut to SetAccessControl with all access disabled
+func (vapp VApp) RemoveAccessControl() error {
+	return vapp.SetAccessControl(&types.ControlAccessParams{IsSharedToEveryone: false})
+}
+
+// IsShared shows whether a vApp is shared or not, regardless of the number of subjects sharing it
+func (vapp VApp) IsShared() bool {
+	settings, err := vapp.GetAccessControl()
+	if err != nil {
+		return false
+	}
+	if settings.IsSharedToEveryone {
+		return true
+	}
+	return settings.AccessSettings != nil
```

Comment:
Here I have to question: We have situation were is shared is false and access Settings == nil means it is still shared?

Reply:
If `IsSharedToEveryone` is false, we have sharing only if settings is not nil and populated.

Reply:
Got it - `IsShared` means more generic - shared to everyone or only to some users, but overall we can say it is shared.